### PR TITLE
Minor: Update docstring on unwrap_cast_in_comparison

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Related to https://github.com/apache/arrow-datafusion/issues/3938

# Rationale for this change

I am working in this code and figured I would encode some of my learning in the docstring

# What changes are included in this PR?

Improve the docstrings 

# Are these changes tested?

No code change
# Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 1]
```diff
@@ -32,25 +32,40 @@ use datafusion_expr::{
...
-/// If this is true, the literal expr will be casted to the data type of expr on the other side, and the result of
-/// binary comparison will be `left_expr comparison_op cast(literal_expr, left_data_type)` or
-/// `cast(literal_expr, right_data_type) comparison_op right_expr`. For better optimization,
-/// the expr of `cast(literal_expr, target_type)` will be precomputed and converted to the new expr `new_literal_expr`
-/// which data type is `target_type`.
-/// If this false, do nothing.
+/// The rule is applied to expressions of the following forms:
+///
+/// 1. `cast(left_expr as data_type) comparison_op literal_expr`
+/// 2. `cast(literal_expr) IN (expr1, expr2, ...)`
```
Comment (viirya) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:48]: I guess that you mean:

```
2. `literal_expr IN (cast(expr1), cast(expr2), ...)`
```

?

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 2]
```diff
@@ -32,25 +32,40 @@ use datafusion_expr::{
 };
 use std::sync::Arc;
 
-/// The rule can be used to the numeric binary comparison with literal expr, like below pattern:
-/// `cast(left_expr as data_type) comparison_op literal_expr` or `literal_expr comparison_op cast(right_expr as data_type)`.
-/// The data type of two sides must be equal, and must be signed numeric type now, and will support more data type later.
+/// [`UnwrapCastInComparison`] Attempts to remove casts from
```
Comment (viirya) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:35]: Attempts -> attempts?

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 3]
```diff
@@ -32,25 +32,40 @@ use datafusion_expr::{
...
-/// the expr of `cast(literal_expr, target_type)` will be precomputed and converted to the new expr `new_literal_expr`
-/// which data type is `target_type`.
-/// If this false, do nothing.
+/// The rule is applied to expressions of the following forms:
+///
+/// 1. `cast(left_expr as data_type) comparison_op literal_expr`
+/// 2. `cast(literal_expr) IN (expr1, expr2, ...)`
+///
+/// If the expression matches one of the forms above, the rule will
+/// ensure the value of `literal` is in within range(min,max) of the
```
Comment (viirya) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:51]: nit: (min,max) -> (min, max)

**Review by viirya — APPROVED** [review]

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs (hunk 3 above)
Comment (alamb) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:51]:
```suggestion
/// ensure the value of `literal` is in within range(min, max) of the
```

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs (hunk 2 above)
Comment (alamb) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:35]:
```suggestion
/// [`UnwrapCastInComparison`] attempts to remove casts from
```

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs (hunk 1 above)
Comment (alamb) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:48]: Yes, that is true. 

```suggestion
/// 2. `literal_expr comparison_op cast(left_expr as data_type)`
/// 3. `cast(literal_expr) IN (expr1, expr2, ...)`
/// 4. `literal_expr IN (cast(expr1) , cast(expr2), ...)`
```

**Review by liukun4515 — APPROVED** [review]
LGTM 
Thanks @alamb

**Pull request merged**